### ISSUE TITLE
--metafile doesn't work with --watch

### ISSUE THREAD

**Issue by JRaspass** [evanw/esbuild#1357]
Unless I'm using it wrong, I can't get the CLI to generate a metafile when under watch mode:

```
$ node_modules/.bin/esbuild --version
0.12.6
```
```
$ ls dist esbuild.json
ls: cannot access 'dist': No such file or directory
ls: cannot access 'esbuild.json': No such file or directory

$ node_modules/.bin/esbuild --metafile=esbuild.json --outdir=dist --watch views/js/hole-ng.js
[watch] build finished, watching for changes...
[watch] build started (change: ".")
[watch] build finished
^C

$ ls dist esbuild.json
ls: cannot access 'esbuild.json': No such file or directory
dist:
hole-ng.js
```
Whereas it works fine without the watch flag:
```
$ node_modules/.bin/esbuild --metafile=esbuild.json --outdir=dist views/js/hole-ng.js

  dist/hole-ng.js  10.2kb

⚡ Done in 10ms

$ ls dist esbuild.json
esbuild.json

dist:
hole-ng.js
```

**Comment by geoffharcourt** [comment]
I think if you want to get the metafile in watch mode you need to do so via a plugin and do it in `onEnd()`. Here's what ours looks like:

```
const metafilePlugin = {
  name: 'metafileWriter',
  setup(build) {
    build.onEnd((result) => {
      if (result.errors.length === 0) {
        fs.writeFileSync(
          path.join(__dirname, 'metafile.json'),
          JSON.stringify(result.metafile),
        );
      }
    });
  },
};
```

**Comment by evanw** [comment]
You're not using it wrong; this is a bug in esbuild. It'll be fixed in the next release. Thanks for reporting this bug.

**Comment by dvodonnell** [comment]
@evanw I believe there might still be an issue here, at least with regenerating the metafile on watch. I was encountering the same issue despite using the latest version of esbuild (0.15.5). I tried it out on a test folder with just a simple single JS file:

`esbuild script/app=a.js --entry-names=[dir]/[name]-[hash] --metafile=asset_meta.json --loader:.png=file --bundle --outdir=assets --watch`

This will successfully create the metafile and the out bundle where it should be. But if I make a change to `a.js`, the new bundle gets created with the new hash, but `asset_meta.json` is not recreated.

I assume this is a bug? If it's intentional behaviour I'm a bit confused. I'm using the above approach to cache-bust in production and have my server serve the right asset file by looking into asset_meta.json, but with this issue I can't keep that flow during development on my local.

**Comment by cyantree** [comment]
I experienced the same with `0.15.10`. Should we create a separate issue for this?

**Comment by evanw** [comment]
Hey, sorry about this. I just independently discovered this as well. It looks like this accidentally regressed when I implemented #1977 in [version 0.14.18](https://github.com/evanw/esbuild/releases/v0.14.18). This bug also happens with the mangle cache feature. I'll fix these bugs and add test coverage for this scenario.

**Comment by DominusKelvin** [comment]
Hey @evanw can you verify if this is actually fixed, I can't get a metafile in watch mode using the JavaScript API either. I also get a type error too 

```
Property 'metafile' does not exist on type 'void | BuildResult<any>'.
  Property 'metafile' does not exist on type 'void'.
```
